### ISSUE TITLE
de-virtualize memtable::partitions_type

### ISSUE THREAD

**Issue by avikivity** [scylladb/scylla#145]
```
class partition_entry {
    bi::set_member_hook<> _link;
    dht::decorated_key _key;
    mutation_partition _p;
    ...
}

...

class memtable final : public enable_lw_shared_from_this<memtable> {
public:
    using partitions_type = bi::set<partition_entry,
        bi::member_hook<partition_entry, bi::set_member_hook<>, &partition_entry::_link>,
        bi::compare<partition_entry::compare>>;
    ...
}
```

because `decorated_key` is polymorphic (through `i_partitioner`), we pay the cost of storing a bytes_type instead of a long, and of using virtual compares instead of long compares.

We could, instead have:

```
template <typename Token>
class devirtualized_decorated_key;

template <typename Token>
class devirtualized_partition_entry;

template <typename Token>
class devirtualized_partitions_type;
```

and store two variations in `memtable` (using a union): one with `Token = int64_t`, one with `Token = dht::token`.

a lookup then pays the cost of just one virtual call (to dispatch to the correct `partitions_type` rather than per compare.

**Comment by avikivity** [comment]
See also #2030.

**Comment by avikivity** [comment]
Fixed (indirectly) by b569d127a09b78fa7f76e08d3493088254d41cf4.
